sparse/coo/common.py

```diff
@@ -609,3 +609,71 @@ def nanreduce(x, method, identity=None, axis=None, keepdims=False, **kwargs):
     """
     arr = _replace_nan(x, method.identity if identity is None else identity)
     return arr.reduce(method, axis, keepdims, **kwargs)
+
+
+def roll(a, shift, axis=None):
+    """
+    Shifts elements of an array along specified axis. Elements that roll beyond
+    the last position are circulated and re-introduced at the first.
+
+    Parameters
+    ----------
+    x : COO
+        Input array
+    shift : int or tuple of ints
+        Number of index positions that elements are shifted. If a tuple is
+        provided, then axis must be a tuple of the same size, and each of the
+        given axes is shifted by the corresponding number. If an int while axis
+        is a tuple of ints, then broadcasting is used so the same shift is
+        applied to all axes.
+    axis : int or tuple of ints, optional
+        Axis or tuple specifying multiple axes. By default, the
+        array is flattened before shifting, after which the original shape is
+        restored.
+
+    Returns
+    -------
+    res : ndarray
+        Output array, with the same shape as a.
+    """
+    from .core import COO, as_coo
+    a = as_coo(a)
+
+    # roll flattened array
+    if axis is None:
+        return roll(a.reshape((-1,)), shift, 0).reshape(a.shape)
+
+    # roll across specified axis
+    else:
+        # parse axis input
+        axis = normalize_axis(axis, a.ndim)
+
+        # make shift iterable
+        if not np.iterable(shift):
+            shift = (shift,)
+
+        # handle broadcasting
+        if np.iterable(axis) and len(shift) == 1:
+            shift = np.full(len(axis), shift)
+
+        # ensure axis is iterable
+        elif not np.iterable(axis):
+            axis = (axis,)
+
+        # check if dimensions are consistent
+        if len(axis) != len(shift):
+            raise ValueError(
+                "If 'shift' is a 1D sequence, "
+                "'axis' must have equal length.")
+
+        if np.asarray(axis).ndim > 1 or np.asarray(shift).ndim > 1:
+            raise ValueError(
```

Comment:
Mandatory: Add test for this `ValueError`, perhaps as a parametrized test in `test_valerr`.

Reply:
I think that should be in there already at line 1585.

Reply:
I see. There are a few things wrong with that. It's not parameterized. And when the top line raises an error, that one is detected and the last line is ignored.

Reply:
You can see here that it isn't covered [here](https://codecov.io/gh/pydata/sparse/compare/eeae480e952328b30bf7e1b85d18318a35332bd8...85bd2cf675492d352ccc89bfbbce2c263f49e532/diff).

Reply:
Got it. Thanks for the patient explanation. Should be fixed now.